### ISSUE TITLE
optional blocks

### ISSUE THREAD

**Issue by glensc** [ccl/dokuwiki-plugin-datatemplate#6]
it would be useful to have optional blocks:

say if some field is not defined, but in template it includes some markup, it would create broken/partial links:

```
| Parent Service          | [[support:@@!parent service@@]] |||
```

in case my page does not define "parent service" i get rendered partial link:

```
| Parent Service          | [[support:]] |||
```

perhaps such markup:

```
#if @@!parent service@@
| Parent Service          | [[support:@@!parent service@@]] |||
#endif
```

**Comment by kasimon** [comment]
The availability of optional blocks would be very nice, but your Problem can be solved with custom aliases too. Define a _support alias in the admin interface with type page and prefix :support: and use the field "parent service_support".

**Comment by glensc** [comment]
your suggestion seems nice, but can't get it to work any way, tried all combinations i could think of
- @@parent_service_support@@
- @@!parent_service_support@@
- [[@@parent_service_support@@]]
- [[@@parent_service@@_support]]
- [[!@@parent_service_support@@]]
- [[!@@parent_service@@_support]]

**Comment by kasimon** [comment]
Sorry for not replying earlier, I got no notification from github. Your have to write the suffix th the datatemplateentry:

---- datatemplateentry ----

## parent_support: whatever

and in the template itself only use @@parent_support@@, dokuwiki will create the link itself.

**Comment by glensc** [comment]
are you sure what are you saying? as afaik formatting from `---- datatemplateentry ----` has no role at all with datatemplate plugin (as it is with data plugin), it is always rendered with template itself, so filling `_support` suffix in datetemplateentry has no point, it needs to be filled in template

i mean did you test what are you suggesting here? i just can't get it work like you say

closest thing i get, is to fill in template:

```
 | Parent Service          | @@!parent@@_support |
```

so it with such dataentry:

```
---- datatemplateentry service ----
template   : tpl_issue6 
Service Name_page  : scms
parent_support : portal
----
```

it can print "portal_support", but as text, not link

**Comment by glensc** [comment]
can you review my wish again and say what can be done here?

**Comment by ccl** [comment]
Sorry for not replying earlier. I did not feel like optional blocks belong in the plugin. I checked if other plugins could maybe provide that possibility (e.g. the macros plugin), but never tried if it really works. However, kasimon is partially right: the plugin supports the _default_ http://www.dokuwiki.org/plugin:data#type_aliases, but only when using @@abc@@ placeholders (not @@!abc@@).

I have the following suggestions and hope that one of them will work for you:
1.  Use two different templates, one with and one without the @@parent@@ placeholder, and then switch between them accordingly.
2.  Use the _page alias:
   - In the datatemplateentry: <pre>parent_page: support:portal</pre>
   - In the template: <pre>| Parent Service          | @@parent@@ |||</pre>
   
   This will result in an empty table cell but no broken links if no parent is defined.
3.  I thought that one could get rid of the whole table row by defining a new type alias as follows:
   - Type: wiki
   - Prefix: <code>| Parent Service | [[support:</code>
   - Postfix <code>]] |||</code>
   
   But it turns out that the current implementation of datatemplate does not support custom column types, and the data plugin does not seem to accept table syntax in a type definition. Hence this solution would require some investigation and programming.

**Comment by glensc** [comment]
yeah, different template is not good way to go, as "parent page" is not only optional field i want to have, i just made sample with one field...

**Comment by ccl** [comment]
What about defining the whole table row inside the datatemplateentry?

<pre>
---- datatemplateentry service ----
template : playground:tpl_issue6 
Service Name_page  : scms
parent : |Parent support|[[support:test]]|
----
</pre>

**Comment by glensc** [comment]
i don't understand, this is supposed to work with current code? can you show both page and template in that case?

**Comment by ccl** [comment]
It should. I have it working here: http://129.194.119.140/wiki/playground:dt_test

**Comment by glensc** [comment]
added some parser can be used like:

```
====== @@service name long@@  ======
| Service                       | [[support:@@!service name@@]] |||
| Description                   | @@service name long@@ |||
{if @@!parent@@}
| Parent Service                | @@!parent@@ |||
{/if}
```

however it has problem with newlines, no matter how i try it will either lose needed newlines, or add extra newlines, so that the table gets broken (extra newline adds new table, or missing newlines make next row being column to previous row)

as it should support combinations of:
- one if between able rows
- several consequtive if statements
